cmd/lakectl/cmd/fs.go

```diff
@@ -409,6 +413,164 @@ var fsRmCmd = &cobra.Command{
 	},
 }
 
+const (
+	fsDownloadCmdMinArgs = 1
+	fsDownloadCmdMaxArgs = 2
+
+	fsDownloadParallelDefault = 6
+)
+
+var fsDownloadCmd = &cobra.Command{
+	Use:   "download <path uri> [<destination path>]",
+	Short: "Download object(s) from a given repository path",
+	Args:  cobra.RangeArgs(fsDownloadCmdMinArgs, fsDownloadCmdMaxArgs),
+	Run: func(cmd *cobra.Command, args []string) {
+		pathURI := MustParsePathURI("path", args[0])
+		flagSet := cmd.Flags()
+		direct := MustBool(flagSet.GetBool("direct"))
+		recursive := MustBool(flagSet.GetBool("recursive"))
+		parallel := MustInt(flagSet.GetInt("parallel"))
+
+		if parallel < 1 {
+			DieFmt("Invalid value for parallel (%d), minimum is 1.\n", parallel)
+		}
+
+		// optional destination directory
+		var dest string
+		if len(args) > 1 {
+			dest = args[1]
+			// verify destination path exists
+			if len(dest) > 0 {
+				pathExists(dest)
+			}
+		}
+
+		// list the files
+		client := getClient()
+		downloadCh := make(chan string)
+		sourcePath := api.StringValue(pathURI.Path)
+
+		// recursive assume source is directory
```

Comment:
Why not support a partial prefix too, the same way we do for la`lakectl fs ls`?

Reply:
1. why do you mean partial prefix?
2. fixed this implementation - had a bug with the prefix.  For single file - just download the file to the destination. For recursive - download all the directory structure from the source to the local destination. Same behaviour of common tools (and aws s3 cp).

Reply:
Partial prefix with `recursive` option should handle all prefixes ("dirs") that start with the prefix. Here's how it looks on `fs ls`:
```
$ ./lakectl fs ls lakefs://myrepo/main/somepre --recursive
object          2022-10-20 14:30:20 -0500 CDT    3.2 kB          somepref/file1
object          2022-10-20 14:30:20 -0500 CDT    3.2 kB          somepref/file2
object          2022-10-20 14:59:49 -0500 CDT    3.2 kB          somepref1/file3
object          2022-10-20 14:59:49 -0500 CDT    3.2 kB          somepref1/file4
```
So I believe one should expect the same objects list to be downloaded by `fs download`, given the same arguments and flags

Reply:
Good point - both commands listing and copy try to replicate aws s3 behaviour.
When you list the prefix will list the files with the same prefix, as you described.
But, if you try to download these files using `aws s3 cp` with the `--recursive` flag it will not copy non folder structure to a folder structure.
I assume this is by design as we like to keep the folder structure on the bucket/repository on the download folder.

Reply:
Interesting. I can see how it aligns with `aws s3 cp` behavior
Why aren't we using `lakectl fs cp` for download  too?

Reply:
We don't have `cp` command - we had so far `upload` and `cat`. Think we can open a new issue to implement `cp` and deprecate all the rest.

![image](https://user-images.githubusercontent.com/8110/197352539-65c0703f-a68e-401a-8529-0243ea9dd07c.png)